### ISSUE TITLE
Vue + ((<SomeType>e.target).localName) = unexpected token

### ISSUE THREAD

**Issue by 3nuc** [typescript-eslint/typescript-eslint#2731]
- [x] I have tried restarting my IDE and the issue persists. (occurs with npx eslint also)
- [x] I have updated to the latest version of the packages (4.6.1-alpha.0)
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

See https://github.com/3nuc/repro or just run
```sh
git clone https://github.com/3nuc/repro.git && cd repro && npm i && npm start
```
eslintrc.js
```js
module.exports = {
  extends: [
    '@vue/typescript/recommended',
  ],
};

```
code:
```TS
<template>
  <div id="app">hello</div>
</template>

<script lang="ts">
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class App extends Vue {
  created(): void {
    document.addEventListener('click', (e) => {
      //comment me out to fix the unexpected token error
      if ((<HTMLInputElement>e.target).localName) console.log('not ok')

      //below versions dont cause issues
      //@ts-ignore
      if (e.target.localName) console.log('ok');
      if ((e.target as HTMLInputElement).localName) console.log('ok');
    });
  }
}
</script>
```

tsconfig.json
```json
{
  "compilerOptions": {
    "strict": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    },
  },
  "exclude": ["node_modules"]
}
```

**Expected Result**

`npx eslint -- src/App.vue`  should return `No lint errors found!`

**Actual Result**

`npx eslint -- src/App.vue` returns

```
error: Parsing error: Unexpected token. Did you mean `{'}'}` or `&rbrace;`? at src/App.vue:20:4:
  18 |       if (e.target.localName) console.log('ok');
  19 |       if ((e.target as HTMLInputElement).localName) console.log('ok');
> 20 |     });
     |    ^
  21 |   }
  22 | }
  23 | </script>
```

**Additional Info**

eslint did not throw an exception. the --debug log is here: https://github.com/3nuc/repro/blob/master/running-npm-start.log

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.6.1-alpha.0` |
| `TypeScript`                | `4.0.5` |
| `ESLint`                    | `7.12.1` |
| `node`                      | `12.8.4` |

I was thinking it's a vue-eslint-parser issue but it only parses the template and delegates parsing <script> to @typescript-eslint/parser anyways

**Comment by bradzacher** [comment]
We don't parse vue. That's all handled by external tooling.
This is occurring because your vue file is being parsed as JSX. So this is working as intended.
The problem is likely a configuration issue.

You will get better support from the config repo that you're using.
https://github.com/vuejs/eslint-config-typescript
